routing/payment_lifecycle.go

```diff
@@ -554,11 +553,15 @@ func (p *shardHandler) collectResult(attempt *channeldb.HTLCAttemptInfo) (
 	}
 
 	// Using the created circuit, initialize the error decrypter so we can
-	// parse+decode any failures incurred by this payment within the
-	// switch.
-	errorDecryptor := &htlcswitch.SphinxErrorDecrypter{
-		OnionErrorDecrypter: sphinx.NewOnionErrorDecrypter(circuit),
-	}
+	// parse+decode any failures incurred by this payment within the switch.
+	//
+	// The resolution format to use for the decryption is based on the
+	// instruction that we gave to the first hop.
+	resolutionFormat := attempt.Route.Hops[0].ResolutionFormat
+
+	errorDecryptor := htlcswitch.NewSphinxErrorDecrypter(
+		circuit, resolutionFormat,
```

Comment:
One case to consider here is that if for some reason one of the nodes had to return `InvalidOnionPayload`, that node wasn't able to read the resolution format. Instead if will return an old-style error that we're not decoding properly. Not sure if we care because it should only happen in case of a bug?